File: src/test/java/nl/wernerkroneman/Drawy/ModelEditor/ParseTreeTest.java
package nl.wernerkroneman.Drawy.ModelEditor;

import nl.wernerkroneman.Drawy.ModelEditor.ParseTree;
import org.junit.Assert;
import org.junit.Test;

public class ParseTreeTest {

    @Test
    public void parseTest() {

        // Parse a tree
        String parserOutput = "Let VB ROOT\n"
                + " +-- be VB ccomp\n"
                + "     +-- there EX expl\n"
                + "     +-- cubes NNS nsubj\n"
                + "         +-- 15 CD num";

        ParseTree tree = new ParseTree(
                parserOutput);

        // Check whether the root word was parsed correctly
        Assert.assertEquals("Let", tree.getRootWord().getRootWord());

        Assert.assertEquals("ROOT", tree.getRootWord().getRole());

        Assert.assertEquals("VB", tree.getRootWord().getNature());

        // Randomly check tree structure
        Assert.assertEquals(1, tree.getRootWord().getChildren().size());

        Assert.assertEquals(2, tree.getRootWord().getChildren().get(0).getChildren().size());

        // Check correct parsing of one of the children
        Assert.assertEquals("cubes", tree.getRootWord().getChildren().get(0).getChildren().get(1).getRootWord());

        Assert.assertEquals("nsubj", tree.getRootWord().getChildren().get(0).getChildren().get(1).getRole());

        Assert.assertEquals("NNS", tree.getRootWord().getChildren().get(0).getChildren().get(1).getNature());

        Assert.assertEquals(parserOutput, tree.toString());
    }

    @Test
    public void parseTestOnlyRoot() {

        ParseTree tree = new ParseTree("Let VB ROOT");

        Assert.assertEquals("Let", tree.getRootWord().getRootWord());

        Assert.assertEquals("ROOT", tree.getRootWord().getRole());

        Assert.assertEquals("VB", tree.getRootWord().getNature());

        Assert.assertEquals(0, tree.getRootWord().getChildren().size());

    }

}

File: src/test/java/nl/wernerkroneman/Drawy/ModelEditor/ParseTreeTest.kt
/*
 * Copyright (c) 2017 Werner Kroneman
 *
 * This file is part of DrawyMcDrawface.
 *
 * DrawyMcDrawface is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DrawyMcDrawface is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DrawyMcDrawface.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.wernerkroneman.Drawy.ModelEditor

import org.junit.Assert
import org.junit.Test

class ParseTreeTest {

    @Test
    fun parseTest() {

        // Parse a tree
        val parserOutput = """
        |Let VB ROOT
        | +-- be VB ccomp
        |     +-- there EX expl
        |     +-- cubes NNS nsubj
        |         +-- 15 CD num
        """.trimMargin()

        val phrase = parsePhraseTree(parserOutput)

        // Check whether the root word was parsed correctly
        Assert.assertEquals("Let", phrase.rootWord)

        Assert.assertEquals("ROOT", phrase.role)

        Assert.assertEquals("VB", phrase.nature)

        // Randomly check tree structure
        Assert.assertEquals(1, phrase.children.size.toLong())

        Assert.assertEquals(2, phrase.children[0].children.size.toLong())

        // Check correct parsing of one of the children
        Assert.assertEquals("cubes", phrase.children[0].children[1].rootWord)

        Assert.assertEquals("nsubj", phrase.children[0].children[1].role)

        Assert.assertEquals("NNS", phrase.children[0].children[1].nature)
    }

    @Test
    fun parseTestOnlyRoot() {

        val phrase = parsePhraseTree("Let VB ROOT")

        Assert.assertEquals("Let", phrase.rootWord)

        Assert.assertEquals("ROOT", phrase.role)

        Assert.assertEquals("VB", phrase.nature)

        Assert.assertEquals(0, phrase.children.size.toLong())

    }

}